File: client/Classes/Basic/NetManager.cpp
//
//  NetManager.h
//  cocos2dx
//
//  Created by Cocos2d on 12-8-9.
//  Copyright (c) 2012年 厦门雅基软件有限公司. All rights reserved.
//

#include "NetManager.h"
#include "BasicFunction.h"
#include "CCNetwork.h"

USING_NS_CC_NETWORK;
using namespace std;

NetManager* NetManager::m_pInstance = NULL;
NetManager::Release NetManager::Garbo;

USING_NS_CC_NETWORK;
using namespace std;

bool NetManager::init(void)
{
//	m_mapError[-1] = "通讯错误或者返回的json数据内容错误";

    m_strToken = "1";
    m_strUrl = "http://localhost:22222/game/";
    m_nIndex = 1;
	return true;
}

bool NetManager::send(ModeRequestType modEnum, DoRequestType doEnum, cocos2d::SEL_CallFuncND selector, cocos2d::CCObject *rec, const char* requestData)
{
    CCNetwork::sharedNetwork()->sendNetPackage(GenerateUrl(doEnum), 1, GeneratePost(modEnum, doEnum, requestData).c_str(), selector, rec);
    
    return true;
}

bool NetManager::sendEx(ModeRequestType modEnum, DoRequestType doEnum, cocos2d::SEL_CallFuncND selector, cocos2d::CCObject *rec, const char* format,...)
{
	va_list   vlist;
	va_start(vlist,format); 
	char temp[1024]="";
	char* param = NULL;
	vsnprintf(temp,1024,format,vlist);
	va_end(vlist);
	if(strlen(temp) > 0)
		param = temp;
    
	return send(modEnum, doEnum, selector, rec, param);
}

string NetManager::GeneratePost(ModeRequestType modEnum, DoRequestType doEnum,const char* requestData)
{
	char temp[10240];
	if (requestData == NULL || strlen(requestData) == 0)
	{
		sprintf(temp, "{\"header\":{\"token\": \"%s\", \"index\": %d}, \"meta\":{\"mod\": \"%s\", \"do\": \"%s\", \"ver\":%d, \"in\":{}}}", m_strToken.c_str(), m_nIndex++, g_modNames[modEnum].c_str(), g_doNames[doEnum].c_str(), g_doVersion[doEnum]);
	} 
	else
	{
		sprintf(temp, "{\"header\":{\"token\": \"%s\", \"index\": %d}, \"meta\":{\"mod\": \"%s\", \"do\": \"%s\", \"ver\":%d,  \"in\":{%s}}}", m_strToken.c_str(), m_nIndex++, g_modNames[modEnum].c_str(), g_doNames[doEnum].c_str(), g_doVersion[doEnum], requestData);
	}
        
	return string(temp);
}

string NetManager::GenerateUrl(DoRequestType doEnum)
{
    return m_strUrl + g_url[doEnum];
}

string NetManager::GetErrorInfo(int errorCode)
{
	map<int, string>::iterator tempIter;
	tempIter = m_mapError.find(errorCode);
	if (tempIter != m_mapError.end())
	{
		return (*tempIter).second;
	}
	
	return "无效的错误代码：" + ConvertToString(errorCode);
}

File: server/src/browser/client/Classes/Basic/NetManager.js
var lc = lc = lc || {};

g_sharedNetManager = null;

lc.NetManager = cc.Class.extend({
    m_iToken:3,
    m_strURL:"http://localhost:22222/game/",
    m_iIndex:1,
    _init:function(){

    },
    getPlatfrom:function(){
        //get current platfrom,return the platfrom of a string;
        return "desktop"
    },
    createRequest:function(){
        var request;
        var platfrom = this.getPlatfrom();
        if(platfrom == lc.g_platfromNames[lc.modePlatfromType.kModeDesktop]){
            request = new lc.CustomXMLHTTPRequest();
        }else{
            // new js binding object for ccnetwork;
        }

        return request;
    },
    //target指向回调的作用域
    sendRequest:function(modEnum,doEnum,requestData,target,responseCallback,errorCallback){
        var request = this.createRequest();
        var url = this.generatePostURL(modEnum,doEnum);
        var parameter = this.generatePostJsonData(modEnum,doEnum,requestData);
        if(parameter == null){
            //send a get request;
        }else{
            //send a post request;
            request.sendRequest(url,parameter,target,responseCallback,errorCallback);
        }
    },
    generatePostJsonData:function(modEnum,doEnum,requestData){
        var jsonText;
        if(requestData == null || requestData == ""){
            jsonText = "{\"header\":{\"token\": \"" + this.m_iToken + "\", \"index\": " + (this.m_iIndex++) + "}, \"meta\":{\"mod\": \"" + g_modNames[modEnum] + "\", \"do\": \"" + g_doNames[doEnum] + "\", \"ver\":" + g_doVersion[doEnum] + ", \"in\":{}}}";
        }else{
            jsonText = "{\"header\":{\"token\": \"" + this.m_iToken + "\", \"index\": " + (this.m_iIndex++) + "}, \"meta\":{\"mod\": \"" + g_modNames[modEnum] + "\", \"do\": \"" + g_doNames[doEnum] + "\", \"ver\":" + g_doVersion[doEnum] + ", \"in\":{" + requestData + "}}}";
        }
        return jsonText;
    },
    generatePostURL:function(modEnum,doEnum){
        return this.m_strURL + g_modNames[modEnum]+"/"+g_doNames[doEnum]+"/";
    }
});

lc.NetManager.sharedNetManager = function(){
    if(g_sharedNetManager == null){
        g_sharedNetManager = new lc.NetManager();
        g_sharedNetManager._init();
    }
    return g_sharedNetManager;
};


